choppa/nextstrain.py

```diff
@@ -0,0 +1,441 @@
+    data_path = "_".join(path.split("/"))
+    return (
+        f"https://data.nextstrain.org/{data_path}.json",
+        nextstrain_metadata[virus]["URL"],
+    )
+
+
+def fetch_nextstrain_json(url):
+    # Make a GET request to the specified URL
+    response = requests.get(url)
+
+    # Raise an exception if the request failed
+    response.raise_for_status()
+
+    # Parse the JSON data and return it
+    return response.json()
+
+
+def fetch_nextstrain_json_mers_cov():
+    with open(MERS_GENOME_TREE_JSON) as f:
+        return json.load(f)
+
+
+def fetch_mers_cov_sequence():
+    """Returns a dict with sequences belonging to MERS-CoV based on https://raw.githubusercontent.com/nextstrain/mers/refs/heads/master/config/mers_reference.gb"""
+    root_sequence_json = {}
+
+    with open(MERS_GENOME_ROOT_SEQUENCE) as handle:
+        # use the somewhat convoluted BioPython GenBank parses to get per-gene protein sequences
+        for record in GenBank.parse(handle):
+            # see https://biopython.org/docs/1.76/api/Bio.GenBank.Record.html?highlight=qualifier#Bio.GenBank.Record.Qualifier
+            for feat in record.features:
+                gene = seq = None
+                if len(feat.qualifiers) == 1:
+                    continue
+                for qual in feat.qualifiers:
+                    if qual.key == "/gene=":
+                        gene = qual.value.replace('"', "")
+                    elif qual.key == "/translation=":
+                        seq = qual.value.replace('"', "")
+                if gene and seq:
+                    root_sequence_json[gene] = seq
+
+    # use Bio.Seq parse genbank to get the 'nuc' entry for the dict
+    gb_entries = SeqIO.parse(MERS_GENOME_ROOT_SEQUENCE, "genbank")
+    root_sequence_json["nuc"] = str(
+        [entry for entry in gb_entries][0].seq
+    )  # always only a single item in the iterator
+    return root_sequence_json
+
+
+def fetch_nextstrain_root_sequence(url, MERS_COV=False):
+    if MERS_COV:  # we've defined this one ourselves, see metadata for notes
+        return fetch_mers_cov_sequence()
+
+    # Header to request the root sequence data
+    headers = {"Accept": "application/vnd.nextstrain.dataset.root-sequence+json"}
+    try:
+        # Make a GET request to the specified URL
+        response = requests.get(url, headers=headers)
+        # Raise an exception if the status code is not 2xx
+        response.raise_for_status()
+
+        return response.json()
+    except requests.exceptions.HTTPError as http_err:
+        # If a 404 error occurs, return None so we can handle it later
+        if response.status_code == 404:
+            return None
+        else:
+            raise http_err
+
+
+def nextstrain_json_to_tree(json_dict, root=True, parent_cumulative_branch_length=None):
+    """
+    Follow the exact same logic that the NextStrain backend follows when calculating per-residue mutation frequencies.
+    By John Huddleston.
```

Comment:
Do we have a link to the source code?

Reply:
here: https://github.com/nextstrain/auspice/blob/e6e4d1a67be4648839f5aff43cb13ce30b23aaa6/src/util/entropy.js#L45-L134, I'll insert.